src/CodeGen_MLIR.cpp

```diff
@@ -0,0 +1,541 @@
+}
+
+void CodeGen_MLIR::Visitor::visit(const GE *op) {
+    value = CompareOpHelper<mlir::arith::CmpIOp, mlir::arith::CmpIPredicate::sge, mlir::arith::CmpIOp, mlir::arith::CmpIPredicate::uge, mlir::arith::CmpFOp, mlir::arith::CmpFPredicate::OGE>(op->type, op->a, op->b);
+}
+
+void CodeGen_MLIR::Visitor::visit(const And *op) {
+    value = builder.create<mlir::arith::AndIOp>(codegen(op->a), codegen(op->b));
+}
+
+void CodeGen_MLIR::Visitor::visit(const Or *op) {
+    value = builder.create<mlir::arith::OrIOp>(codegen(op->a), codegen(op->b));
+}
+
+void CodeGen_MLIR::Visitor::visit(const Not *op) {
+    value = builder.create<mlir::arith::XOrIOp>(
+        codegen(op->a), builder.create<mlir::arith::ConstantOp>(builder.getIntegerAttr(builder.getI1Type(), 1)));
+}
+
+void CodeGen_MLIR::Visitor::visit(const Select *op) {
+    value = builder.create<mlir::arith::SelectOp>(codegen(op->condition),
+                                                  codegen(op->true_value),
+                                                  codegen(op->false_value));
+}
+
+void CodeGen_MLIR::Visitor::visit(const Load *op) {
+    mlir::Value buffer = sym_get(op->name);
+    mlir::Type type = mlir_type_of(op->type);
+    mlir::Value index;
+    if (op->type.is_scalar()) {
+        index = codegen(op->index);
+    } else if (Expr ramp_base = strided_ramp_base(op->index); ramp_base.defined()) {
+        index = codegen(ramp_base);
+    } else {
+        internal_error << "Unsupported Load: " << Expr(op);
+    }
+
+    index = builder.create<mlir::arith::IndexCastOp>(builder.getIndexType(), index);
+    if (op->type.is_scalar()) {
+        value = builder.create<mlir::memref::LoadOp>(type, buffer, mlir::ValueRange{index});
+    } else {
+        value = builder.create<mlir::vector::LoadOp>(type, buffer, mlir::ValueRange{index});
+    }
+}
+
+void CodeGen_MLIR::Visitor::visit(const Ramp *op) {
+    mlir::Value base = codegen(op->base);
+    mlir::Value stride = codegen(op->stride);
+    mlir::Type elementType = mlir_type_of(op->base.type());
+    mlir::VectorType vectorType = mlir::VectorType::get(op->lanes, elementType);
+
+    mlir::SmallVector<mlir::Attribute> indicesAttrs(op->lanes);
+    for (int i = 0; i < op->lanes; i++)
+        indicesAttrs[i] = mlir::IntegerAttr::get(elementType, i);
+
+    mlir::DenseElementsAttr indicesDenseAttr = mlir::DenseElementsAttr::get(vectorType, indicesAttrs);
+    mlir::Value indicesConst = builder.create<mlir::arith::ConstantOp>(indicesDenseAttr);
+    mlir::Value splatStride = builder.create<mlir::vector::SplatOp>(vectorType, stride);
+    mlir::Value offsets = builder.create<mlir::arith::MulIOp>(splatStride, indicesConst);
+    mlir::Value splatBase = builder.create<mlir::vector::SplatOp>(vectorType, base);
+    value = builder.create<mlir::arith::AddIOp>(splatBase, offsets);
```

Comment:
Many `Ramp`s that are generated in Halide have constant stride. I am totally unaware of any of the internals of MLIR, but it seems that generating constants instead of MulIOps seems beneficial in that case. Maybe MLIR further optimizes it's IR and this comment is pointless... I don't know.